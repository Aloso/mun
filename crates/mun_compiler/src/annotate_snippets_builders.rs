//! This file contains a helpful builders for structs from [`annotate-snippets`] crate.
//!
//! [`annotate-snippets`]: https://docs.rs/annotate-snippets/0.6.1/annotate_snippets/

use annotate_snippets::snippet::{Annotation, AnnotationType, Slice, Snippet, SourceAnnotation};
use mun_hir::line_index::LineIndex;

use unicode_segmentation::UnicodeSegmentation;

pub struct SnippetBuilder {
    snippet: Snippet,
}

impl Default for SnippetBuilder {
    fn default() -> Self {
        SnippetBuilder {
            snippet: Snippet {
                title: None,
                footer: vec![],
                slices: vec![],
            },
        }
    }
}

impl SnippetBuilder {
    pub fn new() -> SnippetBuilder {
        SnippetBuilder::default()
    }
    pub fn title(mut self, title: Annotation) -> SnippetBuilder {
        self.snippet.title = Some(title);
        self
    }
    pub fn footer(mut self, footer: Annotation) -> SnippetBuilder {
        self.snippet.footer.push(footer);
        self
    }
    pub fn slice(mut self, slice: Slice) -> SnippetBuilder {
        self.snippet.slices.push(slice);
        self
    }
    pub fn build(self) -> Snippet {
        self.snippet
    }
}

pub struct SliceBuilder {
    slice: Slice,
}

impl SliceBuilder {
    pub fn new(fold: bool) -> SliceBuilder {
        SliceBuilder {
            slice: Slice {
                source: String::new(),
                line_start: 0,
                origin: None,
                annotations: Vec::new(),
                fold,
            },
        }
    }

    pub fn origin(mut self, relative_file_path: String) -> SliceBuilder {
        self.slice.origin = Some(relative_file_path);
        self
    }

    pub fn source_annotation(
        mut self,
        range: (usize, usize),
        label: String,
        source_annotation_type: AnnotationType,
    ) -> SliceBuilder {
        self.slice.annotations.push(SourceAnnotation {
            range,
            label,
            annotation_type: source_annotation_type,
        });
        self
    }

    pub fn build(
        mut self,
        source_text: &str,
        source_text_len: usize,
        line_index: &LineIndex,
    ) -> Slice {
        // Variable for storing first and last line of the needed source code part
        let mut fl_lines: Option<(u32, u32)> = None;

        // Finding borders of source code part that include all highlight ranges
        for annotation in &self.slice.annotations {
            if let Some(range) = fl_lines {
                fl_lines = Some((
                    line_index
                        .line_col(range.0.into())
                        .line
                        .min(line_index.line_col((annotation.range.0 as u32).into()).line),
                    line_index
                        .line_col(range.1.into())
                        .line
                        .max(line_index.line_col((annotation.range.1 as u32).into()).line),
                ));
            } else {
                fl_lines = Some((
                    line_index.line_col((annotation.range.0 as u32).into()).line,
                    line_index.line_col((annotation.range.1 as u32).into()).line,
                ));
            }
        }

        if let Some(fl_lines) = fl_lines {
            self.slice.line_start = fl_lines.0 as usize + 1;
            let first_line_offset = line_index.line_offset(fl_lines.0);

            // Cutting needed part from source code
            self.slice.source = line_index
                .text_part(fl_lines.0, fl_lines.1, source_text, source_text_len)
                .unwrap()
                .to_string();

            // Recalculating every annotation range by taking in account cutted off source text and unicode graphemes
            for annotation in self.slice.annotations.iter_mut() {
                annotation.range = (
                    UnicodeSegmentation::graphemes(
                        &self.slice.source[0..(annotation.range.0 as usize - first_line_offset)],
                        true,
                    )
                    .count(),
                    annotation.range.0 as usize - first_line_offset
                        + UnicodeSegmentation::graphemes(
                            &self.slice.source[annotation.range.0 as usize - first_line_offset
                                ..(annotation.range.1 as usize - first_line_offset)],
                            true,
                        )
                        .count()
                        + (fl_lines.1 - fl_lines.0) as usize, // that line is a temporary fix for annotate-snippets issue number 24
                );
            }
        }
        self.slice
    }
}

pub struct AnnotationBuilder {
    annotation: Annotation,
}

impl AnnotationBuilder {
    pub fn new(annotation_type: AnnotationType) -> AnnotationBuilder {
        AnnotationBuilder {
            annotation: Annotation {
                id: None,
                label: None,
                annotation_type,
            },
        }
    }

    pub fn id(mut self, id: String) -> AnnotationBuilder {
        self.annotation.id = Some(id);
        self
    }

    pub fn label(mut self, label: String) -> AnnotationBuilder {
        self.annotation.label = Some(label);
        self
    }

    pub fn build(self) -> Annotation {
        self.annotation
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn annotation_builder_snapshot() {
        insta::assert_debug_snapshot!(AnnotationBuilder::new(AnnotationType::Note)
            .id("1".to_string())
            .label("test annotation".to_string())
            .build());
    }
    #[test]
    fn slice_builder_snapshot() {
        let source_code = "fn foo():float{\n48\n}";
        let line_index: LineIndex = LineIndex::new(source_code);

        insta::assert_debug_snapshot!(SliceBuilder::new(true)
            .origin("/tmp/usr/test.mun".to_string())
            .source_annotation(
                (14, 20),
                "test source annotation".to_string(),
                AnnotationType::Note
            )
            .build(source_code, source_code.len(), &line_index));
    }
    #[test]
    fn snippet_builder_snapshot() {
        let title = AnnotationBuilder::new(AnnotationType::Note)
            .id("1".to_string())
            .label("test annotation".to_string())
            .build();
        let footer = AnnotationBuilder::new(AnnotationType::Warning)
            .id("2".to_string())
            .label("test annotation".to_string())
            .build();

        let source_code = "fn foo():float{\n48\n}";
        let line_index: LineIndex = LineIndex::new(source_code);

        let slice = SliceBuilder::new(true)
            .origin("/tmp/usr/test.mun".to_string())
            .source_annotation(
                (14, 20),
                "test source annotation".to_string(),
                AnnotationType::Note,
            )
            .build(source_code, source_code.len(), &line_index);

        insta::assert_debug_snapshot!(SnippetBuilder::new()
            .title(title)
            .footer(footer)
            .slice(slice)
            .build());
    }
}
