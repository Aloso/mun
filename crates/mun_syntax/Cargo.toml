[package]
name = "mun_syntax"
version = "0.2.0"
authors = ["The Mun Team <team@mun-lang.org>"]
edition = "2018"
homepage = "https://mun-lang.org"
repository = "https://github.com/mun-lang/mun"
license = "MIT OR Apache-2.0"
description = "Parsing functionality for the Mun programming language"

[dependencies]
rowan = "0.6.1"
text_unit = { version = "0.1.6", features = ["serde"] }
smol_str = { version = "0.1.12", features = ["serde"] }
unicode-xid = "0.1.0"
drop_bomb = "0.1.4"
mun_errors = { path = "../mun_errors"}

[dev-dependencies]
insta = "0.12.0"
