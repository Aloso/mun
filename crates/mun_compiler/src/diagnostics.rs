use mun_hir::diagnostics::{Diagnostic as HirDiagnostic, DiagnosticSink};
use mun_hir::{FileId, HirDatabase, HirDisplay, Module};
use mun_syntax::{ast, AstNode, Parse, SourceFile, SyntaxKind, SyntaxNodePtr, TextRange};

use std::cell::RefCell;

use crate::annotate_snippets_builders::{AnnotationBuilder, SliceBuilder, SnippetBuilder};

use annotate_snippets::snippet::{AnnotationType, Snippet};

fn text_range_to_tuple(text_range: TextRange) -> (usize, usize) {
    (text_range.start().to_usize(), text_range.end().to_usize())
}

fn syntax_node_ptr_location(
    syntax_node_ptr: SyntaxNodePtr,
    parse: &Parse<SourceFile>,
) -> TextRange {
    match syntax_node_ptr.kind() {
        SyntaxKind::FUNCTION_DEF => {
            ast::FunctionDef::cast(syntax_node_ptr.to_node(parse.tree().syntax()))
                .map(|f| f.signature_range())
                .unwrap_or_else(|| syntax_node_ptr.range())
        }
        SyntaxKind::STRUCT_DEF => {
            ast::StructDef::cast(syntax_node_ptr.to_node(parse.tree().syntax()))
                .map(|s| s.signature_range())
                .unwrap_or_else(|| syntax_node_ptr.range())
        }
        _ => syntax_node_ptr.range(),
    }
}

/// Constructs diagnostic messages for the given file.
pub fn diagnostics(db: &impl HirDatabase, file_id: FileId) -> Vec<Snippet> {
    let parse = db.parse(file_id);

    let mut result = Vec::new();

    let source_code = db.file_text(file_id).to_string().replace("\t", " ");
    let source_code_len = source_code.len();

    let relative_file_path = db.file_relative_path(file_id).display().to_string();

    let line_index = db.line_index(file_id);

    result.extend(parse.errors().iter().map(|err| {
        let mut snippet = SnippetBuilder::new()
            .title(
                AnnotationBuilder::new(AnnotationType::Error)
                    .label("syntax error".to_string())
                    .build(),
            )
            .slice(
                SliceBuilder::new(true)
                    .origin(relative_file_path.clone())
                    .source_annotation(
                        (
                            err.location().offset().to_usize(),
                            err.location().end_offset().to_usize(),
                        ),
                        err.to_string(),
                        AnnotationType::Error,
                    )
                    .build(&source_code, source_code_len, &line_index),
            )
            .build();
        // Add one to right range to make highlighting range here visible on output
        snippet.slices[0].annotations[0].range.1 += 1;

        snippet
    }));

    let result = RefCell::new(result);
    let mut sink = DiagnosticSink::new(|d| {
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label("unexpected error".to_string())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(d.highlight_range()),
                            d.message(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::UnresolvedValue, _>(|d| {
        let unresolved_value = d.expr.to_node(&parse.tree().syntax()).text().to_string();

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!(
                            "cannot find value `{}` in this scope",
                            unresolved_value
                        ))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(d.highlight_range()),
                            "not found in this scope".to_string(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::UnresolvedType, _>(|d| {
        let unresolved_type = d
            .type_ref
            .to_node(&parse.syntax_node())
            .syntax()
            .text()
            .to_string();

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!(
                            "cannot find type `{}` in this scope",
                            unresolved_type
                        ))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(d.highlight_range()),
                            "not found in this scope".to_string(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::ExpectedFunction, _>(|d| {
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(d.highlight_range()),
                            format!("expected function, found `{}`", d.found.display(db)),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::MismatchedType, _>(|d| {
        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(d.highlight_range()),
                            format!(
                                "expected `{}`, found `{}`",
                                d.expected.display(db),
                                d.found.display(db)
                            ),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::DuplicateDefinition, _>(|d| {
        let first_definition_location = syntax_node_ptr_location(d.first_definition, &parse);
        let definition_location = syntax_node_ptr_location(d.definition, &parse);

        let duplication_object_type = if matches!(d.first_definition.kind(), SyntaxKind::STRUCT_DEF)
            && matches!(d.definition.kind(), SyntaxKind::STRUCT_DEF)
        {
            "type"
        } else {
            "value"
        };

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(d.message())
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        // First definition
                        .source_annotation(
                            text_range_to_tuple(first_definition_location),
                            format!(
                                "previous definition of the {} `{}` here",
                                duplication_object_type, d.name
                            ),
                            AnnotationType::Warning,
                        )
                        // Second definition
                        .source_annotation(
                            text_range_to_tuple(definition_location),
                            format!("`{}` redefined here", d.name),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .footer(
                    AnnotationBuilder::new(AnnotationType::Note)
                        .label(format!(
                            "`{}` must be defined only once in the {} namespace of this module",
                            d.name, duplication_object_type
                        ))
                        .build(),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::PossiblyUninitializedVariable, _>(|d| {
        let variable_name = d.pat.to_node(&parse.syntax_node()).text();

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!("{}: `{}`", d.message(), variable_name))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(d.highlight_range()),
                            format!("use of possibly-uninitialized `{}`", variable_name),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    })
    .on::<mun_hir::diagnostics::AccessUnknownField, _>(|d| {
        let location = ast::FieldExpr::cast(d.expr.to_node(&parse.syntax_node()))
            .map(|f| f.field_range())
            .unwrap_or_else(|| d.highlight_range());

        result.borrow_mut().push(
            SnippetBuilder::new()
                .title(
                    AnnotationBuilder::new(AnnotationType::Error)
                        .label(format!(
                            "no field `{}` on type `{}`",
                            d.name,
                            d.receiver_ty.display(db),
                        ))
                        .build(),
                )
                .slice(
                    SliceBuilder::new(true)
                        .origin(relative_file_path.clone())
                        .source_annotation(
                            text_range_to_tuple(location),
                            "unknown field".to_string(),
                            AnnotationType::Error,
                        )
                        .build(&source_code, source_code_len, &line_index),
                )
                .build(),
        );
    });

    Module::from(file_id).diagnostics(db, &mut sink);

    drop(sink);

    result.into_inner()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{Color, Config, Driver, PathOrInline, RelativePathBuf};

    #[test]
    fn test_text_range_to_tuple() {
        let text_range = TextRange::from_to(3.into(), 5.into());
        assert_eq!(text_range_to_tuple(text_range), (3, 5));
    }

    /// Compile passed source code and return all compilation errors
    fn compilation_errors(source_code: &str) -> String {
        let config = Config {
            color: Color::Disable,
            ..Config::default()
        };

        let input = PathOrInline::Inline {
            rel_path: RelativePathBuf::from("main.mun"),
            contents: source_code.to_owned(),
        };

        let (driver, _) = Driver::with_file(config, input).unwrap();

        let mut compilation_errors = Vec::<u8>::new();

        let _ = driver.emit_diagnostics(&mut compilation_errors).unwrap();

        String::from_utf8(compilation_errors).unwrap()
    }

    #[test]
    fn test_syntax_error() {
        insta::assert_display_snapshot!(compilation_errors("\n\nfn main(\n struct Foo\n"));
    }

    #[test]
    fn test_unresolved_value_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet b = a;\n\nlet d = c;\n}"
        ));
    }

    #[test]
    fn test_unresolved_type_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a = Foo{};\n\nlet b = Bar{};\n}"
        ));
    }

    #[test]
    fn test_expected_function_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a = Foo();\n\nlet b = Bar();\n}"
        ));
    }

    #[test]
    fn test_mismatched_type_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a: float = false;\n\nlet b: bool = 22;\n}"
        ));
    }

    #[test]
    fn test_duplicate_definition_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn foo(){}\n\nfn foo(){}\n\nstruct Bar;\n\nstruct Bar;\n\nfn BAZ(){}\n\nstruct BAZ;"
        ));
    }

    #[test]
    fn test_possibly_uninitialized_variable_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nfn main() {\nlet a;\nif 5>6 {\na = 5\n}\nlet b = a;\n}"
        ));
    }

    #[test]
    fn test_access_unknown_field_error() {
        insta::assert_display_snapshot!(compilation_errors(
            "\n\nstruct Foo {\ni: bool\n}\n\nfn main() {\nlet a = Foo { i: false };\nlet b = a.t;\n}"
        ));
    }
}
