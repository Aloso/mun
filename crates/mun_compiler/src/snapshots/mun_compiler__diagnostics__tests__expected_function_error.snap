---
source: crates/mun_compiler/src/diagnostics.rs
expression: "compilation_errors(\"\\n\\nfn main() {\\nlet a = Foo();\\n\\nlet b = Bar();\\n}\")"
---
error: cannot find value `Foo` in this scope
 --> main.mun:4:8
  |
4 | let a = Foo();
  |         ^^^ not found in this scope
  |
error: expected function type
 --> main.mun:4:8
  |
4 | let a = Foo();
  |         ^^^ expected function, found `{unknown}`
  |
error: cannot find value `Bar` in this scope
 --> main.mun:6:8
  |
6 | let b = Bar();
  |         ^^^ not found in this scope
  |
error: expected function type
 --> main.mun:6:8
  |
6 | let b = Bar();
  |         ^^^ expected function, found `{unknown}`
  |

